linked_list.py

```diff
@@ -0,0 +1,348 @@
+
+# Defines a node in the singly linked list
+class Node(object):
+    def __init__(self, value=None, next_node=None):
+        self.__data = None  # protected
+        self.next_node = next_node
+        self.__set_data_once(value)
+
+    def get_data(self):
+        if self.__data:
+            return self.__data
+        else:
+            return "Undefined"
+
+    def __set_data_once(self, value):
+        if self.__data == None:
+            self.__data = value
+        else:
+            return ("You may not modify data.  Please add a node.")
+
+# Defines the singly linked list
+class LinkedList(object):
+    def __init__(self):
+        self.__head = None # keep the head private. Not accessible outside this class
+
+    def __retrieve_head(self):
+        return self.head
+
+    def __set_head(self, new_value):
+        self.head = new_value
+
+    # method to add a new node with the specific data value in the linked list
+    # insert the new node at the beginning of the linked list
+    def insert(self, data):
+        new_node = Node(value = data)
+        new_node.next_node = self.__head
+        self.__head = new_node
+        print ("inserting {}".format(data))
+
+    # method to find if the linked list contains a node with specified value
+    # returns true if found, false otherwise
+    def search(value):
+        pass
+
+    
+
+    # method to return the max value in the linked list
+    # returns the data value and not the node
+    def find_max(self):
+        if self.__head == None:
+            return "None"
+
+        node = self.__head
+        max = node.get_data()
+        while node.next_node != None:
+            if node.get_data() > max:
+                max = node.get_data()
+            node = node.next_node
+        return(max)
+
+
+
+    # # method to return the min value in the linked list
+    # # returns the data value and not the node
+    def find_min(self):
+        if self.__head == None:
+            return "None"
+
+        node = self.__head
+        min = node.get_data()
+        while node.next_node != None:
+            if node.get_data() < min:
+                min = node.get_data()
+            node = node.next_node
+        return(min)
+    # # method that returns the length of the singly linked list
+    def length(self):
+        node = self.__head
+        counter = 0
+        while node != None:
+            counter += 1
+            node = node.next_node
+        return(counter)
+
+
+
+
+    # # method to return the value of the nth element from the beginning
+    # # assume indexing starts at 0 while counting to n
+    def find_nth_from_end(self, n):
+        if not self.__head:
+            return
+
+        curr_node = self.__head
+        curr_node_plus_n = self.__head.next_node
+
+        # initialize node_a_plus_n to n elements ahead of counter a
+        for i in range(n):
+            if curr_node_plus_n:

```

Comment:
Help me understand python: is this always checking for the second node i.e. the node following head (`self.__head.next_node`)? I'm unclear on what this does in python - especially since `i` is not getting used at all.